draft-ietf-tls-esni.md

```diff
@@ -446,33 +482,112 @@ to harmonize these to make sure that we maintain key separation.]]
 
 This value is placed in an "encrypted_server_name" extension.
 
-The client MAY either omit the "server_name" extension or provide
-an innocuous dummy one (this is required for technical conformance
-with {{!RFC7540}}; Section 9.2.)
+The client MUST place the value of ESNIKeys.public_name in the "server_name"
+extension. (This is required for technical conformance with {{!RFC7540}};
+Section 9.2.)
+
+If the server negotiates TLS 1.3 or above and provides an
+"encrypted_server_name" extension in EncryptedExtensions, the client
+then processes the extension's "response_type" field:
+
+- If the value is "esni_accept", the client MUST check that the extension's
+  "nonce" field matches ClientESNIInner.nonce and otherwise abort the
+  connection with an "illegal_parameter" alert. The client then proceeds
+  with the connection as usual, verifying the certificate against the desired
+  name.
+
+- If the value is "esni_retry_request", the client proceeds with the handshake,
+  verifying the certificate against ESNIKeys.public_name as described in
+  {{verify-public-name}}. If the handshake completes successfully, the client
+  MUST abort the connection with an "esni_required" alert.
+
+  The client then processes the "retry_keys" field from the server's
+  "encrypted_server_name" extension. If one of the values used a version known
+  to the client, the client SHOULD retry the handshake with a new transport
+  connection, using that value to encrypt the SNI. If no value is applicable,
+  the client SHOULD retry with ESNI disabled.
```

Comment:
"If none of the values provided in "retry_keys" contains a version supported by the client, the client MUST consider the server to have disabled ESNI."

What the client does in response to an authenticated signal that disables ESNI is a matter for client policy, so the SHOULD is inappropriate.

BTW, this would argue for having a third option: esni_disabled.

Reply:
Oops. Missed this one somehow. (GitHub's UI is... not the best).

> "If none of the values provided in "retry_keys" contains a version supported by the client, the client MUST consider the server to have disabled ESNI."

I tweaked the wording a bit based on the first part. It seemed better to adjust the surrounding text too. Re the second part, I'm not sure what "[MUST consider](https://tools.ietf.org/html/rfc6919#section-2)" would mean as a normative requirement...

> What the client does in response to an authenticated signal that disables ESNI is a matter for client policy, so the SHOULD is inappropriate.

The text around unechoed extensions also uses SHOULD. It's true that there isn't a single "hook point" for client policy around the disabling bit. It sounds like you might prefer a more "abstracted" text where something first outputs "secure disable", or "secure replace" and then there's policy on that.

I haven't done that yet, in part because it's late on a Friday, but also I think we're well into wordsmithing weeds now. :-) I don't know what the right documentation organization would be for that. Also it's getting pretty long. Thoughts?

> BTW, this would argue for having a third option: esni_disabled.

It's already there by way of not echoing the extension. I think a separate explicit version would only make sense if there are clients who:
* are happy with `esni_retry_requested`
* would be happy with `esni_disabled`
* are *not* happy with the unechoed signal.

That preference is not entirely implausible because not echoing the extension is an implicit signal repurposing existing server behavior. However, that's also kind of the point, since it allows safe rollback. If server deployments cannot rely on that, the cost of deploying ESNI goes up dramatically.

ESNI, even more so than most TLS improvements, really wants a wide deployment, so I think it's good for the document to have a strong opinion here.

(Specialized clients and specialized servers, of course, would probably ignore disable signals, but they would ignore `esni_disabled` too, so that's more about the SHOULD. But it seems to me that's covered by "there may exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course".)

Reply:
Er, hit comment too soon. Will upload the corresponding git commit shortly.

Reply:
(Uploaded.)